src/resolve/resolved-dns-cache.c

```diff
@@ -533,6 +546,10 @@ static int dns_cache_put_negative(
         assert(key);
         assert(owner_address);
 
+        /* Don't store the negative responses if the StaleRetentionSec is greater than zero. */
+        if (stale_retention_usec > 0 && rcode != DNS_RCODE_NXDOMAIN)
```

Comment:
this looks wrong to me. this shouldn't be filtered when storing negative answers in the cache, but when using them from the cache. After all negative caching should always be used when TTL allows it, even if stale retention is on.

also DNS_RCODE_SUCCESS should be treated like DNS_RCODE_NXDOMAIN here, i.e. for NODATA cached entries.

Reply:
Removed the filter condition while putting the negative cache